### PR TITLE
Parse clocking declarations

### PR DESCRIPTION
**Pull request by evanbryantmiller**

### CODE REVIEW

File: tools/syntax.txt [hunk 1]
```diff
@@ -542,6 +542,38 @@ token keyword
...
+token globalOrDefault
+token clocking
+token blockName
+token at
+ParenthesizedEventExpressionSyntax event
+token eventIdentifier
+token semi
+list<ClockingItemSyntax> items
+token endClocking
+token colon
```
Comment (MikePopoloski) [inline tools/syntax.txt]: Should use NamedBlockClauseSyntax so that analysis code for that can be unified.

File: source/Parser.cpp [hunk 2]
```diff
@@ -442,6 +442,13 @@ MemberSyntax* Parser::parseMember() {
             return alloc.emplace<EmptyMemberSyntax>(attributes, nullptr, consume());
         case TokenKind::PropertyKeyword:
             return parsePropertyDeclaration(attributes);
+        case TokenKind::GlobalKeyword:
+        case TokenKind::DefaultKeyword:
+            if (peek(1).kind != TokenKind::ClockingKeyword) {
+                break;
+            }
```
Comment (MikePopoloski) [inline source/Parser.cpp]: Don't rely on fallthrough, it's confusing.

File: source/Parser.cpp [hunk 3]
```diff
@@ -1907,6 +1914,125 @@ ParameterDeclarationSyntax* Parser::parseParameterPort() {
     return alloc.emplace<ParameterDeclarationSyntax>(Token(), type, parseOneVariableDeclarator());
 }
 
+ClockingSkewSyntax* Parser::parseClockingSkew() {
+    Token edge, hash;
+    ExpressionSyntax* value = nullptr;
+    if (peek().kind == TokenKind::EdgeKeyword || peek().kind == TokenKind::PosEdgeKeyword || peek().kind == TokenKind::NegEdgeKeyword) {
```
Comment (MikePopoloski) [inline source/Parser.cpp]: There's a peek() method that takes a TokenKind and returns a bool which would make this easier.

File: source/Parser.cpp [hunk 4]
```diff
@@ -1907,6 +1914,125 @@ ParameterDeclarationSyntax* Parser::parseParameterPort() {
...
+        edge = consume();
+    }
+    if (peek().kind == TokenKind::Hash) {
+        hash = consume();
+        switch(peek().kind) {
+            case TokenKind::OpenParenthesis:
+            {
```
Comment (MikePopoloski) [inline source/Parser.cpp]: Braces on previous line.

File: source/Parser.cpp [hunk 5]
```diff
@@ -1907,6 +1914,125 @@ ParameterDeclarationSyntax* Parser::parseParameterPort() {
...
+    Token semi = expect(TokenKind::Semicolon);
+    SmallVectorSized<ClockingItemSyntax*, 4> buffer;
+    if (globalOrDefault.kind != TokenKind::GlobalKeyword) {
+        while(peek().kind != TokenKind::EndClockingKeyword && peek().kind != TokenKind::EndOfFile) {
+            Token defaultKeyword, inputKeyword, outputKeyword;
+            ClockingDirectionSyntax* direction;
+            ClockingSkewSyntax* inputSkew = nullptr, *outputSkew = nullptr;
+            MemberSyntax* declaration;
+            switch(peek().kind) {
+                case TokenKind::DefaultKeyword:
```
Comment (MikePopoloski) [inline source/Parser.cpp]: The use of fallthrough in this switch statement is pretty confusing.

File: source/Parser.cpp [hunk 6]
```diff
@@ -1907,6 +1914,125 @@ ParameterDeclarationSyntax* Parser::parseParameterPort() {
...
+            auto closeParen = expect(TokenKind::CloseParenthesis);
+            event = alloc.emplace<ParenthesizedEventExpressionSyntax>(openParen, innerExpr, closeParen);
+        } else {
+            eventIdentifier = expect(TokenKind::Identifier);
+        }
+    }
+    Token semi = expect(TokenKind::Semicolon);
+    SmallVectorSized<ClockingItemSyntax*, 4> buffer;
+    if (globalOrDefault.kind != TokenKind::GlobalKeyword) {
+        while(peek().kind != TokenKind::EndClockingKeyword && peek().kind != TokenKind::EndOfFile) {
```
Comment (MikePopoloski) [inline source/Parser.cpp]: You have to be careful with this kind of loop; if the endclocking keyword is missing then it will loop forever.

Comment (evanbryantmiller) [inline source/Parser.cpp]: What should I do about this then? Should I include other tokens to check against here?

Comment (MikePopoloski) [inline source/Parser.cpp]: You can look at functions like parseMemberList (in fact, you might be able to just use that directly). It does a lot of work to gracefully handle errors.

Comment (MikePopoloski) [inline source/Parser.cpp]: parseBlockItems might be a better example in that it also checks isEndKeyword(), which means if the user forgot the endclocking but has some other end token in their file (like endmodule or whatever) we won't just suck up every token.

**Pull request merged**